Fixes #7889: Add Puppet Modules to the database.

ehelms opened the pull request:
Opening this up for initial review as this should cover all of the workflows involved with puppet modules. This includes:

- [x] Top level puppet modules page with details
- [x] Updates to puppet module references throughout the UI
- [x] Puppet sync and module upload
- [x] Addition of puppet modules to content views
- [x] Publishing and promoting a content view containing puppet modules
- [x] Viewing the details of a content view version with puppet modules
- [x] Import upgrade rake task
- [x] Migrations and indexes to add puppet module AR and associations
- [x] Updated tests

jlsherrill commented on app/controllers/katello/api/v2/content_views_controller.rb:
```diff
@@ -93,21 +93,15 @@ def show
     param :name, String, :desc => N_("module name to restrict modules for"), :required => false
     def available_puppet_modules
-      current_ids = @view.content_view_puppet_modules.map(&:uuid).reject { |p| p.nil? }
+      current_uuids = @view.content_view_puppet_modules.map(&:uuid).reject { |p| p.nil? }
```
`@view.content_view_puppet_modules.where("uuid is not NULL").pluck(:uuid)`  would be better

jlsherrill commented on app/controllers/katello/api/v2/content_views_controller.rb:
```diff
@@ -93,21 +93,15 @@ def show
-      search_filters = [{ :terms => { :repoids => repo_ids }}]
+      query = PuppetModule.in_repositories(repositories)
+      query = query.where(:name => params[:name]) if params[:name]
+      query = query.where("#{PuppetModule.table_name}.uuid NOT in (?)", current_uuids.join(',')) if current_uuids.present?
```
no need to do the current_uuids.join here,  AR knows how to handle an array of ids

jlsherrill commented on app/controllers/katello/api/v2/content_views_controller.rb:
```diff
@@ -115,17 +109,14 @@ def available_puppet_modules
 
-      options = {:filters  => search_filters}
+      modules = PuppetModule.in_repositories(repos)
+      modules = modules.where('name NOT in (?)', current_names.join(',')) if current_names.present?
```
same comment as above

ehelms commented:
Latest update includes:

 * merging of repository_content_controller and repository_db_content_controller mixin
 * removal of glue/elasticsearch/repository.rb -- index command was moved to glue/pulp/repo.rb
 * updates to all failing tests
 * rebase on master
 * addressing comments up to this point

I am still planning to do a round of new tests to ensure coverage.

cfouant commented on app/models/katello/content_view_version.rb:
```diff
@@ -78,6 +78,12 @@ def self.for_version(version)
     end
 
+    def self.with_puppet_module(puppet_module)
+      debugger
```
I think this needs removing (maybe rubocop missed?)

ehelms replied:
Nope! rubocop is nice and red at the moment.

cfouant commented:
APJ for me, but definitely deferring to @jlsherrill for a fuller testing than I am capable of. Thanks @ehelms!

ehelms commented:
I am currently working through some updates and fixes based on failures and adding more tests which is leading to some minor code re-work.

ehelms commented:
Updated and fixed tests -- everything is now green. Note here that I also merged repository_db_content_controller and repository_content_controller into one.

jlsherrill commented on app/models/katello/content_view_puppet_module.rb:
```diff
@@ -17,18 +17,21 @@ class ContentViewPuppetModule < Katello::Model
+        puppet_module = PuppetModule.latest_module(
+          self.name,
+          self.author,
+          self.content_view.puppet_repos
```
I think this should be   'self.organization.library.repositories.puppet_type' and not the repos in the content view

ehelms replied:
I might be missing something, but thats what the `puppet_repos` function does? https://github.com/Katello/katello/blob/master/app/models/katello/content_view.rb#L177

jlsherrill replied:
oh good call, i misread that.

jlsherrill commented on app/models/katello/puppet_module.rb:
```diff
@@ -26,5 +56,32 @@ def self.parse_metadata(filepath)
+
+    def self.with_identifiers(ids)
+      self.where(:id => ids)
+    end
```
did we not want to support uuid and id here?

ehelms replied:
I think the method didn't exist when I first wrote this, I'll remove it in favor of the PulpContentUnit method.

jlsherrill commented:
I don't see a puppet_module model test.  I'd like to see that added.

ehelms commented:
Updated with more puppet_module model tests

jlsherrill commented:
Seeing an error when publsihing a content view:
```
NoMethodError: undefined method `with_uuid' for Katello::PuppetModule:Class
Backtrace:
/home/vagrant/git/katello/app/models/katello/content_view_puppet_environment.rb:90:in `index_content'
/home/vagrant/git/katello/app/lib/actions/pulp/content_view_puppet_environment/index_content.rb:12:in `run'
/home/vagrant/.rvm/gems/ruby-1.9.3-p448@foreman/gems/dynflow-0.8.4/lib/dynflow/action.rb:506:in `block (3 levels) in execute_run'
```

jlsherrill commented:
Now getting:

```
2015-09-04T12:28:55 [app] [E] RuntimeError: Content type not supported for removal
 | /home/vagrant/git/katello/app/models/katello/repository.rb:547:in `removable_unit_association'
 | /home/vagrant/git/katello/app/models/katello/repository.rb:482:in `units_for_removal'
```

also don't forget to update the SUBJECT

ehelms commented:
Updated to fix broken tests and removal and updated the subject.

jlsherrill commented:
ACK
